hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/ContainerSafeModeRule.java

```diff
@@ -48,101 +48,117 @@
-    this(ruleName, eventQueue, conf, containerManager.getContainers(), containerManager, manager);
-  }
 
-  public ContainerSafeModeRule(String ruleName, EventQueue eventQueue,
-             ConfigurationSource conf,
-             List<ContainerInfo> containers,
-             ContainerManager containerManager, SCMSafeModeManager manager) {
+  public ContainerSafeModeRule(final String ruleName,
+                               final EventQueue eventQueue,
+                               final ConfigurationSource conf,
+                               final ContainerManager containerManager,
+                               final SCMSafeModeManager manager) {
     super(manager, ruleName, eventQueue);
+    this.safeModeCutoff = getSafeModeCutoff(conf);
     this.containerManager = containerManager;
-    safeModeCutoff = conf.getDouble(
-        HddsConfigKeys.HDDS_SCM_SAFEMODE_THRESHOLD_PCT,
-        HddsConfigKeys.HDDS_SCM_SAFEMODE_THRESHOLD_PCT_DEFAULT);
-
-    Preconditions.checkArgument(
-        (safeModeCutoff >= 0.0 && safeModeCutoff <= 1.0),
-        HddsConfigKeys.HDDS_SCM_SAFEMODE_THRESHOLD_PCT  +
-            " value should be >= 0.0 and <= 1.0");
+    this.ratisContainers = new HashSet<>();
+    this.ecContainers = new HashSet<>();
+    this.ecContainerDNsMap = new ConcurrentHashMap<>();
+    initializeRule();
+  }
 
-    ratisContainers = new HashSet<>();
-    ecContainers = new HashSet<>();
-    ecContainerDNsMap = new ConcurrentHashMap<>();
 
-    initializeRule(containers);
+  private static double getSafeModeCutoff(ConfigurationSource conf) {
+    final double cutoff = conf.getDouble(HDDS_SCM_SAFEMODE_THRESHOLD_PCT,
+        HDDS_SCM_SAFEMODE_THRESHOLD_PCT_DEFAULT);
+    Preconditions.checkArgument((cutoff >= 0.0 && cutoff <= 1.0),
+        HDDS_SCM_SAFEMODE_THRESHOLD_PCT  +
+            " value should be >= 0.0 and <= 1.0");
+    return cutoff;
   }
 
-
   @Override
   protected TypedEvent<NodeRegistrationContainerReport> getEventType() {
     return SCMEvents.CONTAINER_REGISTRATION_REPORT;
   }
 
   @Override
   protected synchronized boolean validate() {
-    return (getCurrentContainerThreshold() >= safeModeCutoff) &&
-        (getCurrentECContainerThreshold() >= safeModeCutoff);
+    if (validateBasedOnReportProcessing()) {
+      return (getCurrentContainerThreshold() >= safeModeCutoff) &&
+          (getCurrentECContainerThreshold() >= safeModeCutoff);
+    }
+
+    // TODO: Split ContainerSafeModeRule into RatisContainerSafeModeRule and
+    //   ECContainerSafeModeRule
+    final List<ContainerInfo> containers = containerManager.getContainers(
+        ReplicationType.RATIS);
+
+    final Set<ContainerID> containerIDs = containers.parallelStream()
+        .filter(this::isClosed).map(ContainerInfo::containerID)
+        .collect(Collectors.toSet());
+
+    final long missingContainerCount = containerIDs.parallelStream()
+        .filter(this::isMissing).count();
+
+    return missingContainerCount == 0;
```

Comment:
I think we should avoid using `parallelStream` without comparing performance to regular one.

Also, we don't need to materialize the results of the first stream:

```suggestion
    return containers.stream()
        .filter(this::isClosed)
        .map(ContainerInfo::containerID)
        .noneMatch(this::isMissing);
```

Reply:
The reason for materializing the `ContainerIDs` and calculating `missingContainerCount` is to use this info in later PRs for reporting purpose.

This can be done when required. I will update the suggested change.